src/server/qtquick/private/wquickforeigntoplevelmanagerv1.cpp

```diff
@@ -0,0 +1,116 @@
+// Copyright (C) 2023 Dingyuan Zhang <lxz@mkacg.com>.
+// SPDX-License-Identifier: Apache-2.0 OR LGPL-3.0-only OR GPL-2.0-only OR GPL-3.0-only
+
+#include "wglobal.h"
+#include "woutput.h"
+#include "wquickforeigntoplevelmanagerv1_p.h"
+#include "wxdgsurface.h"
+
+#include <qwforeigntoplevelhandlev1.h>
+#include <qwxdgshell.h>
+
+#include <map>
+#include <memory>
+
+using QW_NAMESPACE::QWForeignToplevelHandleV1;
+using QW_NAMESPACE::QWForeignToplevelManagerV1;
+
+WAYLIB_SERVER_BEGIN_NAMESPACE
+
+class WQuickForeignToplevelManagerV1Private : public WObjectPrivate {
+  public:
+    WQuickForeignToplevelManagerV1Private(WQuickForeignToplevelManagerV1 *qq)
+        : WObjectPrivate(qq) {}
+
+    void updateSurfaceData(WXdgSurface *surface) {
+        auto handle = surfaces[surface];
+
+        QObject::connect(surface, &WXdgSurface::titleChanged, q_func(),
+                         [=] { handle->setTitle(surface->title().toUtf8()); });
+
+        QObject::connect(surface, &WXdgSurface::appIdChanged, q_func(),
+                         [=] { handle->setAppId(surface->appId().toUtf8()); });
+
+        QObject::connect(surface, &WXdgSurface::requestMinimize, q_func(),
+                         [=] { handle->setMinimized(surface->isMinimized()); });
+
+        QObject::connect(surface, &WXdgSurface::requestMaximize, q_func(),
+                         [=] { handle->setMaximized(surface->isMaximized()); });
+
+        QObject::connect(
+            surface, &WXdgSurface::requestFullscreen, q_func(),
+            [=] { handle->setFullScreen(surface->isFullScreen()); });
+
+        QObject::connect(surface, &WXdgSurface::activateChanged, q_func(),
+                         [=] { handle->setActivated(surface->isActivated()); });
+
+        QObject::connect(
+            surface, &WXdgSurface::parentSurfaceChanged, q_func(),
+            [this, surface, handle] {
+                auto find = std::find_if(
+                    surfaces.begin(), surfaces.end(),
+                    [surface](auto pair) { return pair.first == surface; });
+
+                handle->setParent(find != surfaces.end() ? find->second.get()
+                                                         : nullptr);
+            });
+    }
+
+    void add(WXdgSurface *surface) {
+        auto handle = std::make_shared<QWForeignToplevelHandleV1>(
+            QWForeignToplevelHandleV1::create(manager));
+        surfaces.insert({surface, handle});
+        updateSurfaceData(surface);
+    }
+
+    void remove(WXdgSurface *surface) { surfaces.erase(surface); }
+
+    void surfaceOutputEnter(WXdgSurface *surface, WOutput *output) {
+        auto handle = surfaces[surface];
```

Comment:
Should I check that the surface still exists?

Reply:
You can assert it.